socks5.py

```diff
@@ -52,24 +68,28 @@
         else:
             iftypes['cell'].append(iface)
 
+
     if iftypes['bridge']:
         iface = iftypes['bridge'][0]
-        print("Assuming proxy will be accessed over hotspot (%s) at %s" %
-              (iface.name, iface.addr.address))
+        initial_output = "Assuming proxy will be accessed over hotspot (%s) at %s\n" % (iface.name, iface.addr.address)
         PROXY_HOST = iface.addr.address
     elif iftypes['en']:
         iface = iftypes['en'][0]
-        print("Assuming proxy will be accessed over WiFi (%s) at %s" %
-              (iface.name, iface.addr.address))
+        initial_output += "Assuming proxy will be accessed over WiFi (%s) at %s\n" % (iface.name, iface.addr.address)
         PROXY_HOST = iface.addr.address
     else:
-        print('Warning: could not get WiFi address; assuming %s' % PROXY_HOST)
+        initial_output += 'Warning: could not get WiFi address; assuming %s\n' % PROXY_HOST
 
     if iftypes['cell']:
-        iface = iftypes['cell'][0]
-        print("Will connect to servers over interface %s at %s" %
-              (iface.name, iface.addr.address))
-        CONNECT_HOST = iface.addr.address
+        iface_ipv4 = next((iface for iface in iftypes['cell'] if iface.addr.family == socket.AF_INET), None)
+        iface_ipv6 = next((iface for iface in iftypes['cell'] if iface.addr.family == socket.AF_INET6), None)
+        if iface_ipv4:
+            initial_output += "Will connect to servers over interface %s at %s\n" % (iface_ipv4.name, iface_ipv4.addr.address)
```

Comment:
maybe mention "Will connect to IPv4 servers...", "Will connect to IPv6 servers..."

Reply:
Will have in the next commit. Your comment here actually pointed out to me that for some reason the 'if iface_ipv6' line does not print currently, even though connections do flow over IPv6. I'll debug this further.

Reply:
@nneonneo Maybe you can help me here. There's a bug that I am having a problem resolving. The "if iface_ipv6" code never executes. I believe this is only working at all via good luck of this code:
    def connect_to_address(self, address, port, family, log_tag, timeout=None):
        try:
            remote = socket.socket(family, socket.SOCK_STREAM)
            remote.settimeout(timeout)
            if family == socket.AF_INET and "ipv4" in CONNECT_HOST:
                if CONNECT_HOST["ipv4"]:
                    remote.bind((CONNECT_HOST["ipv4"], 0))
            elif family == socket.AF_INET6 and "ipv6" in CONNECT_HOST:
                if CONNECT_HOST["ipv6"]:
                    remote.bind((CONNECT_HOST["ipv6"], 0))
            remote.connect((address, port)) <-- This code will still execute even if there is no value for CONNECT_HOST["ipv6"], and it will just allow the OS to use whatever IPv6 address it wants to use (which, in the case of my testing, is the only ipv6 address with a valid default gateway, which is the cellular link.

I suspect the bug is in the ifaddrs.py.

I tried changing the code as follows:
def get_sockaddr(sockaddr_p):
    if not sockaddr_p:
        return None
    sa_len, sa_family = cast(sockaddr_p, POINTER(c_ubyte*2)).contents
    if sa_family == socket.AF_INET:
        sin = copy_zerofill(SockaddrIn(), sockaddr_p, sa_len)
        buf = create_string_buffer(16+1)
        addr4 = bytes(sin.sin_addr)
        libc.inet_ntop(sa_family, sin.sin_addr, buf)
        return SocketAddress(sa_family, buf.value.decode())
    elif sa_family == socket.AF_INET6:
        sin6 = copy_zerofill(SockaddrIn6(), sockaddr_p, sa_len)
        buf = create_string_buffer(46+1)
        addr = bytes(sin6.sin6_addr)
        libc.inet_ntop(sa_family, byref(sin6.sin6_addr), buf)
        return SocketAddress(sa_family, buf.value.decode())
    else:
        sa_data = cast(sockaddr_p, POINTER(c_char*sa_len)).contents.raw
        return SocketAddress(sa_family, sa_data)


But it did not fix the issue.

If you would like to help me debug this, you should be able to do it just by configuring a dummy ipv6 address on a mac ... even if the ipv6 address doesn't work, the code should still be able to at least see the address.



Unfortunately, until this issue is fixed, solving the delay in 'happy eyeballs' can't effectively be done. Once this issue is solved, I plan on solving the 'happy eyeballs' delay in two stages. Stage 1 will be to fix it using Futures by having it evaluate if there is an IPv6 address set on the cellular interface, and if there is an IPv6 address on the host we are trying to connect to, and only attempting happy eyeballs if both are true. Perhaps it can also do an initial test when it launches to see if it can get an IPv6 connection to www.google.com and if not, assume that IPv6 doesn't work even if an address is set. 


The second stage of the fix is to move from Futures to async sockets ... which is a much larger refactor of the whole code... and realistically may not be necessary if the 'stage 1' fix works well enough.

Reply:
Scratch that, I think I found the issue ... it's most likely the "TODO IPv6 support someday" check which I misunderstood the purpose for...